axe-testing/src/crawl-page.js

```diff
@@ -0,0 +1,68 @@
+// Crawl and collect URLs for accessibility checks
+import { AxePuppeteer } from '@axe-core/puppeteer';
+
+export async function crawlPage(
+  page,
+  browser,
+  HOMEPAGE_URL,
+  visitedPages,
+  allResults,
+  blacklistUrls,
+) {
+  const currentUrl = page.url();
+  let uniqueUrl = currentUrl;
+
+  // Skip if page has been visited already
+  if (visitedPages.has(currentUrl)) {
+    console.log(`Skipping already visited page: ${currentUrl}`);
+    return;
+  }
+
+  // Mark page as visited
+  visitedPages.add(currentUrl);
+
+  // Skip if the URL is blacklisted
+  if (Array.isArray(blacklistUrls) && blacklistUrls.includes(currentUrl)) {
+    console.log(`Skipping as on blacklist: ${currentUrl}`);
+    return;
+  }
+
+  // Distinguish between the login page and the dashboard with the same URL for the results
+  if (currentUrl === HOMEPAGE_URL) {
+    uniqueUrl += '-post-login';
+  }
+
+  console.log(`Assessing page: ${uniqueUrl}`);
+
+  // Run Axe accessibility checks on the current page
+  const results = await new AxePuppeteer(page).analyze();
+
+  // Add the results to allResults
+  allResults.push({
+    url: uniqueUrl,
+    results,
+  });
+
+  // Get all links on the current page
+  const links = await page.$$eval('a', (anchors) =>
+    anchors.map((anchor) => anchor.href),
+  );
+
+  // Crawl through each link found on the current page
+  for (const link of links) {
+    // Avoid revisiting the same pages or external URLs
+    if (!visitedPages.has(link) && link.startsWith(HOMEPAGE_URL)) {
+      const newPage = await browser.newPage();
+      await newPage.goto(link, { waitUntil: 'networkidle2' });
+      await crawlPage(
+        newPage,
+        browser,
+        HOMEPAGE_URL,
+        visitedPages,
+        allResults,
+        blacklistUrls,
+      ); // Recursively crawl new pages
+      await newPage.close(); // Close new page after crawling
+    }
+  }
```

Comment:
This is good! It doesn't cover everything as you probably know (which is very hard to do with crawling). I'll recap some of what's not covered here, but this is just for reference and future refinement, not for trying to address in _this_ PR.

1) some routes only exist if the right data is present (e.g. links to collection management or upload routes from the home page only exist if the logged in user has collections to manage/upload to)
    - could be addressed by seeding the database with specially crafted data before performing the accessibility scan. It'll be a small maintenance burden, but it's that or try and make a crawler that can correctly fill out and submit valid forms, and then re-crawl the site for dynamic changes/new forms until everything has been reached... which is going to be slower and a nightmare to build, haha.
2) navigation can also occur via button clicks (e.g. when the server response includes a redirect, which we won't generally do, or when client side JS navigates the client based on server response, such as moving the user the the collection management route for a new collection after submitting a collection creation form)
    - off the top of my head, and ideally it stays this way, any non-link navigation leads to a page that is also reachable by link, if the data for it exists. Might be able to ignore this problem :crossed_fingers: 
3) elements that render/respond to user input won't be interacted with by a shallow scan on each route (e.g. owner and uploader email input fields don't exist until the user clicks the button to add new emails)
    - this is also likely not worth trying to fully automate as part of our end-to-end scans, that's too hard a problem. Likewise, maintaining puppeteer scripts to exercise dynamic elements on each page would be a big maintenance burden. I'd prefer to leave interactive elements out of scope in this end-to-end crawler, and instead figure out a unit test (component level) equivalent for more dynamic accessibility checks.

Reply:
Item 1 is the only one I'd prioritize for follow up. Happy to discuss it any time you want!